### PR TITLE
Eliminate usages of getters generated by the wrapper

### PR DESCRIPTION
**Pull request by ice1000**
This depends on #217.

Please feel free to spot any getter calls that I didn't noticed!

### CODE REVIEW

**Comment by nrc** [comment]
Nice! Could you add a generation option to protobuf-build to not generate getters? Then the compiler can tell us if we changed them all (and generate less code :-) ).

**Comment by ice1000** [comment]
@nrc I can do that!

File: src/log_unstable.rs [hunk 1]
```diff
@@ -64,16 +64,13 @@ impl Unstable {
...
     }
 
     /// Returns the last index if it has at least one unstable entry or snapshot.
     pub fn maybe_last_index(&self) -> Option<u64> {
         match self.entries.len() {
-            0 => self
-                .snapshot
-                .as_ref()
-                .map(|snap| snap.get_metadata().get_index()),
+            0 => self.snapshot.as_ref().map(|snap| snap.get_metadata().index),
```
Comment (siddontang) [inline src/log_unstable.rs]: do we still need to call `get_metadata()`?

Comment (ice1000) [reply src/log_unstable.rs]: `.metadata` returns `Option`, `.get_metadata` will return default value if `None`

**Comment by ice1000** [comment]
> Nice! Could you add a generation option to protobuf-build to not generate getters? Then the compiler can tell us if we changed them all (and generate less code :-) ).

This depends on tikv/protobuf-build#7

**Comment by hicqu** [comment]
LGTM. PTAL @nrc .

**Comment by ice1000** [comment]
This depends on another unmerged PR and some extra work is required :joy: it's not the time for LGTM ATM

**Comment by Hoverbear** [comment]
Placing this in DNM while we wait for the improvement. :) Thanks!

File: src/log_unstable.rs (hunk 1 above)
Comment (ice1000) [reply src/log_unstable.rs]: However, seems like there's a `metadata()` function.

Comment (ice1000) [reply src/log_unstable.rs]: No, there isn't.

**Comment by ice1000** [comment]
> Nice! Could you add a generation option to protobuf-build to not generate getters? Then the compiler can tell us if we changed them all (and generate less code :-) ).

Done.

**Comment by ice1000** [comment]
@Hoverbear @nrc @hicqu I've addressed the comments. This PR is no longer DNM, PTAL

File: Cargo.toml [hunk 2]
```diff
@@ -14,7 +14,7 @@ edition = "2018"
 build = "build.rs"
 
 [build-dependencies]
-protobuf-build = "0.4.1"
+protobuf-build = { git = "https://github.com/tikv/protobuf-build", branch = "redesign" }
```
Comment (nrc) [inline Cargo.toml]: Need to wait for the release before we land this PR

File: build.rs [hunk 3]
```diff
@@ -52,7 +53,12 @@ fn main() {
             .map(|m| format!("src/prost/{}.rs", m))
             .collect::<Vec<_>>(),
         "src/prost",
-        GenOpt::All,
+        GenOpt::MUT
+            | GenOpt::TRIVIAL_SET
+            | GenOpt::HAS
+            | GenOpt::TAKE
+            | GenOpt::CLEAR
+            | GenOpt::MESSAGE,
```
Comment (nrc) [inline build.rs]: Nicer to start with `All` and then remove trivial get?

Comment (ice1000) [reply build.rs]: I recommend removing the remaining parts one by one instead of adding the eliminated parts one by one.

File: Cargo.toml (hunk 2 above)
Comment (ice1000) [reply Cargo.toml]: Landed

**Comment by ice1000** [comment]
@hicqu @Hoverbear PTAL

File: src/raft.rs [hunk 4]
```diff
@@ -938,7 +934,7 @@ impl<T: Storage> Raft<T> {
 
     fn num_pending_conf(&self, ents: &[Entry]) -> usize {
         ents.iter()
-            .filter(|e| e.get_entry_type() == EntryType::EntryConfChange)
+            .filter(|e| e.entry_type() == EntryType::EntryConfChange)
```
Comment (Hoverbear) [inline src/raft.rs]: Not possible to remove the getters on these too? Or later work?

Comment (ice1000) [reply src/raft.rs]: Nope. The field is `i32` while we want to get an enum.

This getter is actually generated by prost by default.

**Pull request merged**